library/alloc/src/vec/spec_from_iter_nested.rs

```diff
@@ -23,8 +23,19 @@ where
         let mut vector = match iterator.next() {
             None => return Vec::new(),
             Some(element) => {
-                let (lower, _) = iterator.size_hint();
-                let mut vector = Vec::with_capacity(lower.saturating_add(1));
+                let guess_rest_len = match iterator.size_hint() {
+                    // Trivial cases
+                    (lower, None) => lower,
+                    (lower, Some(upper)) if lower == upper => lower,
+                    // This may prevent overallocation when we ever reallocate memory for Vec.
+                    (lower, Some(upper)) if lower * 2 >= upper => upper,
+                    // There we guess that there is a chance that there is more than `lower` items
+                    (lower, Some(upper)) if upper > lower * 2 => lower * 2,
+                    // Maybe there is possible to do some another estimations
+                    // but let's do some conservative estimate for now.
+                    (lower, _) => lower,
```

Comment:
Won't this case never be hit? `upper > lower * 2` is true whenever `lower * 2 >= upper` is false.

Reply:
Rewritten cases in match.